Implement mod discovery

DaRealTurtyWurty opened the pull request.

DaRealTurtyWurty committed:
Basic mod discovery implementation

TODO:
- SecureJar
- Validation
- Additional URLs
- Dependencies

DaRealTurtyWurty committed:
Changes

DaRealTurtyWurty committed:
modid validation and loading

DaRealTurtyWurty committed:
Update ModDiscovery.java

DaRealTurtyWurty committed:
Update ModDiscovery.java

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+        try {
+            final List<SecureJar> mods = new ArrayList<>();
+
+            Files.walkFileTree(MODS_FOLDER, new SimpleFileVisitor<>() {
```
This will visit jars in subdirectories as well. Please use the overload with a fixed depth of `1`.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    try {
+                        // Read jar
+                        SecureJar secureJar = SecureJar.from(file);
```
```suggestion
                        if (!file.toString().endsWith(".jar")) return FileVisitResult.CONTINUE;
                        SecureJar secureJar = SecureJar.from(file);
```

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+                        mods.add(secureJar);
+
+                        Path metadata = secureJar.getPath("mod.hjson");
+                        InputStream stream;
+                        if (Files.exists(metadata))
+                            stream = Files.newInputStream(metadata);
+                        else
+                            return FileVisitResult.CONTINUE;
```
```suggestion
                        if (Files.notExists(metadata)) return FileVisitResult.CONTINUE;
                        final var data = Files.readString(metadata);
```
And adapt the code to use the string.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+                        JsonObject jsonObject = JsonValue.readHjson(new String(stream.readAllBytes())).asObject();
+
+                        // Get and validate the mandatory mod details
+                        String modid = getString(jsonObject, "modid", file);
```
```suggestion
                        String modid = getString(jsonObject, "modId", file);
```
Json properties should use camelCase.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+
+                        // Get and validate the mandatory mod details
+                        String modid = getString(jsonObject, "modid", file);
+                        lengthCheck("modid", modid, file, 3, 20);
```
```suggestion
                        lengthCheck("modId", modid, file, 3, 20);
```
Same as above. Additionally, consider adding a helper method that returns the property from the json, while also length-checking it.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+                        Path icon = Path.of(iconStr);
+
+                        // Get optional mixin usage (boolean)
+                        boolean usesMixins = jsonObject.getBoolean("usesMixins", false);
```
This seems like unneeded duplication. Just check for the existence of the `mixins` block which will be implemented in the future.

DaRealTurtyWurty replied:
Mixins block? What do you mean? Mixins are done via the mixins config.

Matyrobbrt replied:
I mean,
```json
"mixins": ["mymod.mixins.json"] 
```

Although, thinking about it, the `usesMixins` seems rather useless when you can just check it the mixins configs array is empty. Even then, what is the purpose of the property?

DaRealTurtyWurty replied:
Yea I guess on second thought, it is quite redundant.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+            String[] arr = itemsJson.asArray().values().stream().map(JsonValue::asString).toArray(String[]::new);
+            for (int i = 0; i < arr.length; i++) {
+                String author = arr[i];
+                lengthCheck(name + " " + (i + 1), author, modLoc, 4, 32);
```
The additional credits use the same method, and they may be structured like  `Thanks to SomeVeryLongUsername for some very long feature.`, which will surpass 32 characters. Please add an int parameter for the limit.

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+            items = arr;
+        } else if(itemsJson != null && itemsJson.isString()) {
+            items = new String[] { itemsJson.asString() };
+        } else
+            items = new String[0];
```
```suggestion
        } else {
            items = new String[0];
        }
```

Matyrobbrt commented on src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    private static boolean modidAllows(char c) {
+        return c == '_' || c == '-' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.';
```
Please remove dashes as valid mod ID characters. JMPS does not allow dashes in module names, and the module of the mod jars will have the mod ID as their name. 
```suggestion
        return c == '_' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.';
```

Matyrobbrt commented on src/mixin/java/io/github/brassmc/brassloader/mixin/BootstrapMixin.java:
```diff
@@ -0,0 +1,48 @@
+                Constructor<?> constructor = mainClass.getConstructor();
+                constructor.setAccessible(true);
+                constructor.newInstance();
+            } catch (ClassNotFoundException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] points to an invalid class!");
+            } catch (InvocationTargetException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] threw an exception!");
+            } catch (InstantiationException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] is abstract!");
+            } catch (IllegalAccessException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] is not accessible!");
+            } catch (NoSuchMethodException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] does not contain a no-args constructor!");
+            }
```
Collapse this block into a `catch (Exception)`, the Java exceptions messages are detailed enough.

DaRealTurtyWurty replied:
I prefer to handle this myself rather than the less descriptive default implementations.

Matyrobbrt replied:
In that case, please add the original exception as a parent of the new exception.

Matyrobbrt commented on src/mixin/java/io/github/brassmc/brassloader/mixin/BootstrapMixin.java:
```diff
@@ -0,0 +1,48 @@
+            } catch (NoSuchMethodException exception) {
+                throw new InvalidEntrypointException("Entrypoint in mod [" + modContainer.modid() + "] does not contain a no-args constructor!");
+            }
+            System.out.println("Loaded mod: " + modContainer.modid());
```
Use a logger.

DaRealTurtyWurty replied:
Do we have a logger already or is that something that needs to be added?

Matyrobbrt replied:
No, but consider moving the contents of the init method in a `ModLoader` class, as a package-protected method called using reflection. The logger can be put in that class.

DaRealTurtyWurty committed:
Jar verification

Co-authored-by: matyrobbrt <matyrobbrt@gmail.com>

DaRealTurtyWurty committed:
Cleanup

Co-authored-by: matyrobbrt <matyrobbrt@gmail.com>

DaRealTurtyWurty committed:
Camel Case

Co-authored-by: matyrobbrt <matyrobbrt@gmail.com>

DaRealTurtyWurty committed:
Camel Case

Co-authored-by: matyrobbrt <matyrobbrt@gmail.com>

DaRealTurtyWurty committed:
Disallow '-' in modIDs

Co-authored-by: matyrobbrt <matyrobbrt@gmail.com>

Matyrobbrt commented on src/mixin/java/io/github/brassmc/brassloader/mixin/BootstrapMixin.java:
```diff
@@ -0,0 +1,48 @@
+            ),
+            method = "bootStrap"
+    )
+    private static void init(CallbackInfo callback) {
```
Also, no idea how I missed it before, but brass prefix. 
```suggestion
    private static void brass$loadMods(CallbackInfo callback) {
```

DaRealTurtyWurty committed:
Mod Discovery Cleanup

- Extracted utility methods to class
- Created utility method for getting a string and length checking in one
- Prevent recursive file walking
- Add limit option to `getArrayOrString`
- Remove useless mixins option
- Added logger

DaRealTurtyWurty commented:
I have updated according to all the requested changes
